pkg/kubernetes/scripts/virtual-machines/components/createVmDialog.jsx

```diff
@@ -0,0 +1,264 @@
+
+        this.onResourceChanged = this.onResourceChanged.bind(this);
+
+        this.getResult = this.getResult.bind(this);
+        this.validate = this.validate.bind(this);
+    }
+
+    onResourceChanged(value) {
+        this.setState({ resource: value });
+    }
+
+    validate() {
+        let validation = true;
+
+        let resourceError = null;
+
+        try {
+            if (!this.state.resource) {
+                resourceError = _("VM definition is required.");
+            } else {
+                let res = JSON.parse(this.state.resource);
+                if (res === null || typeof res !== "object") {
+                    resourceError = _("VM definition is not a valid JSON.");
+                } else if (res instanceof Array) {
+                    resourceError = _("VM definition should be an object. Not an array.");
+                }
+            }
+        } catch (e) {
+            resourceError = _("VM definition is not a valid JSON.");
+        }
+
+        this.setState({ resourceError: resourceError });
+        if (resourceError) {
+            validation = false;
+        }
+
+        return validation;
+    }
+
+    getResult() {
+        let result = {
+            resource: null,
+        };
+
+        try {
+            let res = JSON.parse(this.state.resource);
+            if (!(res instanceof Array)) {
+                result.resource = res;
+            }
+        } catch (ignored) {
+            console.warn("getResult() called before validate()");
+        }
+
+        return result;
+    }
+
+    onNewFile(file) {
+        if (file.size > MAX_IMPORT_FILE_SIZE_MIB * 1024 * 1024) {
+            this.setState({
+                resourceError: cockpit.format(_("Only files of size $0 MiB and less are supported"), MAX_IMPORT_FILE_SIZE_MIB),
+            });
+            return;
+        }
+        this.setState({ filename: file.name });
+        let reader = new FileReader();
+        reader.onload = (e) => this.setState({ resource: e.target.result });
+        reader.readAsText(file);
+    }
+
+    onNewFileEvent(e) {
+        if (e.target.files.length > 0) {
+            this.onNewFile(e.target.files[0]);
+        }
+    }
+
+    onDrop(e) {
+        e.preventDefault();
+        this.setState({ isDragging: false });
+        const dt = e.dataTransfer;
+        if (dt.items) {
+            for (let i = 0; i < dt.items.length; i++) {
+                if (dt.items[i].kind === "file") {
+                    this.onNewFile(dt.items[i].getAsFile());
+                    break;
+                }
+            }
+        } else {
+            if (dt.files.length > 0) {
+                this.onNewFile(dt.files[0].getAsFile());
+            }
+        }
+    }
+
+    onDragEnter() {
+        this.setState({ isDragging: true });
+    }
+
+    onDragLeave() {
+        this.setState({ isDragging: false });
+    }
+
+    onDragEnd(e) {
+        // cleanup
+        const dt = e.dataTransfer;
```

Comment:
Is the cleanup necessary? What happens if it's omitted?

Reply:
I guess it is not needed, since every new drag operation starts with new files